### ISSUE TITLE
META: FORTRAN Code inventory

### ISSUE THREAD

**Issue by ilayn** [scipy/scipy#18566]
This issue is meant for a central place for our Fortran (almost exclusively F77) codebase to track per module and if possible remove depending on availability of contributor affinity of the subject matter and fortran-fu. 

List has projected effort required just based on code length hence involves prejudice.

- [ ] scipy.integrate:
  -  [ ] dop Convert: Easy
  - [ ] mach Convert: Easy
  - [ ] odepack : Convert : Hard
  - [ ] quadpack : Convert : Moderate-Hard
- [ ] scipy.interpolate
  - [ ] fitpack : Convert : Moderate-Hard
- [ ] scipy.io
  - [ ] Test Fortran : ? 
- [x] scipy.linalg
  - [x] lu/det : Replaced in #18225, #18358
  - [x] id_dist : Convert: Moderate (replacement in progress)
- [ ] scipy.odr
  - [ ] odrpack : Convert : Moderate
- [ ] scipy.optimize
  - [ ] nnls: Convert : Easy/Moderate
  - [ ] cobyla : Convert : Hard (The new implementation PRIMA is going to replace cobyla)
  - [ ] L-BFGS-B: Convert : Hard
  - [ ] minpack: Convert : Hard (reimplemented by LFortran team)
  - [ ] minpack2: Convert : Easy
  - [ ] slsqp : Convert: Hard
- [ ] scipy.sparse.linalg
  - [ ] ARPACK Convert : Hard
  - [x] iterative Replaced : #18488
  - [ ] PROPACK : Convert : Hard
- [ ] scipy.special
  - [ ] AMOS: Convert : Hard
  - [ ] cdflib: Convert : Hard
  - [ ] mach: Convert : Easy
  - [ ] specfun : Convert: Very Hard

**Comment by steppi** [comment]
`specfun`actually isn't so bad because each function is mostly self contained, and it can be converted bit by bit. I've already made good headway with `hyp2f1` and that's arguably the most difficult function to replace. I'm determined to banish the F77 from special and have time to get back to it now.

I recently took a close look at ODEPACK to review gh-14552. Converting that will be a bear since it consists of large pieces of machinery with sophisticated control logic. I'd be up for helping to replace it with a C, C++, or Cython implementation if we can find someone else with sufficient, time, knowledge, and interest to work on it together with me. I think it would have to wait until the all of the deliverables are met for the SDG @tirthasheshpatel and I have to work on `special` though.

**Comment by j-bowhay** [comment]
I am interested in looking at `dop` over the summer as I am fairly familiar with Hairer's work however my Fortran knowledge is fairly minimal so won't block anyone else's efforts.

**Comment by ilayn** [comment]
You don't need to know Fortran. If you know about the functionality then you can approximate as closely as possible without following the code faithfully. You don't need to jump through the GOTOs but kind of get why and where it jumps. I am basically following the logic and not the code. Because you go quickly insane with Arithmetic IFs and [Computed GOTOs](https://docs.oracle.com/cd/E19957-01/805-4939/6j4m0vn9l/index.html).

**Comment by Kai-Striega** [comment]
I'd also be interested in picking up some of these. I'll take a close look during the week and pick somewhere to get started.

**Comment by steppi** [comment]
I'd be happy to review PRs for replacing old Fortran code. I think I can take a look at gh-18570 this week.

> I am interested in looking at `dop` over the summer as I am fairly familiar with Hairer's work however my Fortran knowledge is fairly minimal so won't block anyone else's efforts.

Picking up the language itself is actually not too hard. You could look through https://web.stanford.edu/class/me200c/tutorial_77/ to get started, and look things up here https://docs.oracle.com/cd/E19957-01/805-4939/index.html as needed as you work through code. Like @ilayn said, it's the unstructured control flow that can make understanding the code so difficult. 

> You don't need to jump through the GOTOs but kind of get why and where it jumps. I am basically following the logic and not the code. Because you go quickly insane with Arithmetic IFs and [Computed GOTOs](https://docs.oracle.com/cd/E19957-01/805-4939/6j4m0vn9l/index.html).

What I tend to do is follow along with a notebook and try to work out flowcharts in an almost mechanical fashion, and then use these to reason about parts of the program. I've found trying to follow along and keep track of the state in ones head like one can do in structured programs is basically impossible.

**Comment by zachjweiner** [comment]
@j-bowhay The `dop` methods already have pure-Python implementations accessible via the "new" interface, `solve_ivp`. LSODA is the only wrapped Fortran method via the new interface. The *vode methods (also from odepack) are the only ones available in the old interface but not the new one, but I'm not sure how much use they get (or "should" get) in practice. 

LSODA does provide unique functionality (automatic stiffness detection and switching between stiff/nonstiff methods), but I don't know that the algorithm itself is especially complex. The stiff and nonstiff methods individually should be straightforward  (pure-Python BDF---LSODA's stiff solver---is already available via `solve_ivp`). Though the stiffness detection and logic to handle switching would be new, from skimming the [algorithm description](https://www.researchgate.net/publication/239878868_Automatic_Selection_of_Methods_for_Solving_Stiff_and_Nonstiff_Systems_of_Ordinary_Differential_Equations) it only seems modestly involved.

But rather than simply directly reimplementing LSODA, I think it would be a better use of time to implement abstracted utilities for method switching/composition (which could then be used to replicate LSODA). DifferentialEquations.jl [does this](https://docs.sciml.ai/DiffEqDocs/stable/solvers/ode_solve/#CompositeAlgorithm) and finds that combinations of other stiff and nonstiff methods can outperform LSODA. Obviously this goes beyond the mission of this issue, but if effort is to be spent on the ODE solvers I don't think it should be to simply directly translate LSODA/odepack. With a proper, modular abstraction, I would expect the individual components (stiff and nonstiff solvers with their own dense output, stiffness detection, and logic for dynamically switching) to be simpler to tackle than the LSODA source would suggest.[^1] (And I don't immediately see a reason these should be implemented in compiled code, at least at first.)

On a semi-related note: a year ago I rewrote `solve_ivp` (and just the Runge-Kutta methods) in Cython to see how much overhead could be mitigated. There were a number open questions/issues to address - just mentioning here if anyone is interested in discussing elsewhere.

[^1]: There is an (apparently inactive) [C rewrite](https://github.com/sdwfrost/liblsoda) of LSODA, which may be easier to reference for anyone digging into it.

**Comment by ilayn** [comment]
I don't know enough solver knowledge to contribute to your analysis. But it sounds like we can indeed have a look at the Cython implementation to see whether we can improve or I don't know what you would like to get out of that exercise. Typically a verbose Cython code can get to wrapped Fortran speed with not so much effort. All depends on the sophistication. 

The main issue we are trying to tackle is to get rid of this dormant unmaintained code which is blocking us to achieve more user friendly and feature rich API. In a way, I come to believe that the code you wrap forces you into its own API, good or bad. And it is impossible to troubleshoot these things. #14807 is one of my favorite bug reports I've encountered here. 

Hence if domain experts think there are better alternatives, then we can switch to them. If they think it's a waste of time, then we deprecate it. Or it is a daunting task to do anything about but we still need the functionality, then a few of us who are crazy enough can sit down and give it a go. After spending last month with this fortran code, I'm simultaneously amazed that the authors provided such robust code which still works and being used by so many people, and lost almost all my respect to scientific computing community at the same time for letting this code untouched for the last 40 years making the same excuses.

**Comment by zachjweiner** [comment]
> But it sounds like we can indeed have a look at the Cython implementation to see whether we can improve or I don't know what you would like to get out of that exercise. Typically a verbose Cython code can get to wrapped Fortran speed with not so much effort. All depends on the sophistication.

I'm happy to share it but I should've been clearer that I think it's independent from the Fortran issue (I was just rewriting the pure-Python routines in Cython). Just wanted to mention for visibility to people thinking about compiled code for IVP solvers. (And indeed I was able to get the overhead to `<< 1 us` but only with an API change, and I hadn't tackled dense output.) But the odepack routines come from compiled code only because odepack is an old, widely-used implementation, not because they intrinsically need to be compiled to be useful/reasonably performant.

> Hence if domain experts think there are better alternatives, then we can switch to them. If they think it's a waste of time, then we deprecate it.

I should've also mentioned that, since a large fraction of the odepack-dependence comes only from the "old" `ode` API, it would probably be best to decide whether to actually, officially deprecate it first. After that, my only point is that I think it would be more worthwhile in the long run to enhance/add to the functionality under the `solve_ivp` umbrella (as I described above) than to painstakingly translate LSODA from Fortran.

> And it is impossible to troubleshoot these things. https://github.com/scipy/scipy/issues/14807 is one of my favorite bug reports I've encountered here. In other words, scientific python still smells like fortran-like API which is a shame if you compare it with other coding domains and frameworks.

Thanks for sharing - would be glad to see the ecosystem move toward being not only more maintainable and robust to such correctness issues, but also more extensible and improvable.

**Comment by j-bowhay** [comment]
Thanks @zachjweiner for the summary. I did wonder what the overlap was with the existing Python code but was yet to put in any effort to investigate. If I understand you correctly, the only barrier to removing the `dop` directory is the legacy ode solvers? If so I will probably focus my efforts elsewhere on something hopefully more impactful.

**Comment by ev-br** [comment]
There is also some Fortran code in stats,  all probably rather low-hanging fruit. 

```
$ ll scipy/stats/*f
-rw-rw-r-- 1 br br 46387 мая  8  2022 scipy/stats/mvndst.f
-rw-rw-r-- 1 br br  2563 ноя 21  2022 scipy/stats/mvn.pyf
-rw-rw-r-- 1 br br  1528 ноя 21  2022 scipy/stats/statlib.pyf
```

**Comment by ev-br** [comment]
This is great @ilayn !
Now that we have a laundry list, I think it would be helpful to add two more classification indicators:
- whether a module causes problems or not. For instance, ODRPACK seems stable and without open issues. So probably low prio.
- alternatives, both external and interternal. For instance, VODE: I don't think it makes much sense to just transcribe the fortran implementation without taking a look at SUNDIALS which is in C++. Also `solve_ivp`, as discussed. (For ode solving specifically, I think it's best to discuss it in a separate issue @zachjweiner --- I think a cythonized solve_ivp might be very welcome).
One other example is QUADPACK. Maybe the best course of action is to assess what's missing in `quad_vec`, port the missing bits; once it's feature complete, we can just deprecate `quad`.

**Comment by ilayn** [comment]
Indeed, I am very much all in for redesigning certain items if the codebase left us quite behind. The DifferentialEquations.jl is an example of how a specialized task force can take things way forward than what this fortran code can offer. So in that sense, I'm all in. One thing I am starting to notice is that, once you understand the codeflow you start to get rather grandiose ideas about how to generalize so better careful with that :) 

Unfortunately, I'm seriously illiterate when it comes to anything that is not linalg or optimize. Hence, I am not sure if I can be of any help in designing the strategy. What I want is to get rid of the usual suspects, `ARPACK`/`PROPACK` and maybe (and by that I mean really maybe) `L-BFGS-B` stuff if I have steam left.

So TL;DR, let's

- replace the ones for which a modern implementation exists, 
- translate the required/industry standard ones,
-  and deprecate the rest. 

If `integrate` work can be consolidated as discussed, then let's do it. It shouldn't have been us doing this, we have to start somewhere anyways.

**Comment by ilayn** [comment]
One thing we have to be careful about stability statement is that though they are stable, they are fantastically old and use outdated ways of doing things in the absence of modern tooling (say fixed lapack bugs, threaded/reentrant C libs, and so on). Core of `id_dist` for example is doing hand coded householder trafos, and qr orthogonalizations which are now almost one-liners. 

Long story short, though they are stable, these code have huge inertia for newcomers to implement things and also carry a mystique that these code are battle-tested hence should not be touched and so on which is demonstrably not true until you actually look at it 😃

**Comment by dschmitz89** [comment]
I had a look into the optimizers. The biggest problem is in my opinion `l-bfgs-b`. There are many `l-bfgs` implementations but no good ones that work with bounds (the `l-bfgs-b` part). For `slsqp`, I see a few possibilitites:

- modern and actively maintained Fortran version [here](https://github.com/jacobwilliams/slsqp): no gotos but Fortran tooling problems still apply
- auto translated version of the F77 code from [nlopt](https://github.com/stevengj/nlopt/blob/master/src/algs/slsqp/slsqp.c): still full of gotos but at least in C and actively maintained. We also rely on nlopt for the `DIRECT` algorithm

I agree with @ilayn 's observation that the cores of these algorithms often look like homegrown linear algebra algorithms or copies of the classic F77 implementations which are nowadays available in LAPACK. Even the modern Fortran code includes again [copies of BLAS routines](https://github.com/jacobwilliams/slsqp/blob/master/src/slsqp_support.f90) and constrained least squares [codes](https://github.com/jacobwilliams/slsqp/blob/master/src/bvls_module.f90). To avoid linking complexity, this probably makes sense for Fortran or C but for scipy ..

**Comment by matthew-brett** [comment]
Just to clarify - does the modern and actively maintained Fortran project present any _new_ problems in tooling?  I think that's the key question.

**Comment by ilayn** [comment]
I don't know if we have any actively maintained new fortran code (maybe PRIMA will be if we manage to link it) so probably we would need to experience it to see if there are new issues.

Folks mention iso_c_binding and ctypes routes but I didn't see any demonstration yet. If we don't need to wrestle with ABI issues and/or linking strangeness, I'm fine with it. Like I mentioned before, this is not a crusade against Fortran, but to get rid of dormant code which happens to be fortran77.

**Comment by ev-br** [comment]
cc https://github.com/scipy/scipy/issues/19079#issuecomment-1682108384 for a C++ port of AMOS, which seems to be license-compatible.

**Comment by ivan-pi** [comment]
A full Python interface for MINPACK is available since one year now: https://github.com/fortran-lang/minpack/tree/main/python

A few tests serving as examples can be found here: https://github.com/fortran-lang/minpack/blob/main/python/minpack/test_library.py

But I guess this doesn't really solve any of the problems you have with F77 which are:
- easily installing a Fortran compiler on all OS's
- the constraints the legacy codes impose when it comes to more feature rich and Pythonic interfaces
- lack of developers interested in learning Fortran
- the fact NumPy defaults to row-major

It appears easier for people to create something new in their own (closed) community than to work across programming languages and tools.

Some time ago I rewrote NNLS with modern Fortran flow-control constructs in an attempt to make it more maintainable and easy to interface. Unfortunately, I cannot live off of refactoring Fortran codes, hence I had no motivation to contribute it to SciPy, despite being a user.

**Comment by inkydragon** [comment]
> blas: Convert: Very Hard (Depends on BLAS/LAPACK lib) A lot of historical wrapper inconsistencies

Maybe try [BLIS](https://github.com/flame/blis) which was written in C.
[Performance](https://github.com/flame/blis/blob/master/docs/Performance.md) (Compare with OpenBLAS, MKL, and Eigen)

> BLIS is written in ISO C99 and available under a [new/modified/3-clause BSD license](http://opensource.org/licenses/BSD-3-Clause). While BLIS exports a [new BLAS-like API](https://github.com/flame/blis/blob/master/docs/BLISTypedAPI.md), it also includes a BLAS compatibility layer which gives application developers access to BLIS implementations via traditional [BLAS routine calls](http://www.netlib.org/lapack/lug/node145.html).

AMD also builds their BLAS library ([AOCL-BLAS](https://www.amd.com/en/developer/aocl/dense.html)) based on BLIS. 
> AOCL-BLAS is developed as a forked version of BLIS

**Comment by ilayn** [comment]
Yes BLIS is in my radar. I need to spare some time and see how we can find a fitting LAPACK for it. In the meantime, Rust folks are also covering quite some distance. I am looking into those. The absolutely last resort is going to be have a performant BLAS-like, and write a small subset of LAPACK with it. But I am really not looking forward to it. 

Julia folks did some really nice progress with [`RecursiveFactorization.jl`](https://juliapackages.com/p/recursivefactorization) and I have something similar in mind but it is even more work then rewriting all Fortran code or so it seems to me.

**Comment by cournape** [comment]
For ARPACK, I wonder how feasible it would be use https://github.com/JuliaLinearAlgebra/ArnoldiMethod.jl as a reference and implement it in cython/etc.

According to https://discourse.julialang.org/t/ann-arnoldimethod-jl-v0-4/110604, it is more stable than ARPACK.

**Comment by rgommers** [comment]
Thanks for sharing @cournape. That looks great - way easier to understand than ARPACK and MIT-licensed.

**Comment by haampie** [comment]
Note that ArnoldiMethod.jl only implements the non-symmetric case, it's lacking Lanczos. I don't know what the state of the art is, but as far as I remember Lanczos needs extra orthogonalization because it's unstable, and with that in place it looks rather similar to the non-symmetric case.

**Comment by ilayn** [comment]
I need some help vetting SUNDIALS for the solvers. Odepack is extra bad F77 code and seems quite pointless to translate if we are going to mostly deprecate them. I'm a bit running out of steam with these last ones stuff so if there is a replacement let's switch to them. But I don't have enough experience to judge so I'd appreciate if there is some detailed look into alternatives.

**Comment by j-bowhay** [comment]
This is a pretty accurate description of the current landscape 

https://www.stochasticlifestyle.com/comparison-differential-equation-solver-suites-matlab-r-julia-python-c-fortran/

**Comment by ilayn** [comment]
Yes, that is still current, but does not help us if a new implementation is not possible on our side. I certainly don't have the expertise to write a new robust implementation of solvers in my limited time. So I need help if something can replace the existing odepack drop-in style.

**Comment by j-bowhay** [comment]
The only piece of fortran left in the non-legacy interface is LSODE. The article suggests CVODE from SUNDIALS would be a worthy replacement to this.

**Comment by ilayn** [comment]
So does that mean SUNDIALS + (LSODE translation) would cover all our API surface?

**Comment by j-bowhay** [comment]
I think it depends if people still care about the legacy `odeint` +`ode` interface, this is where ODEPACK is mostly used. Many of these methods have already been translated to python and are exposed via `solve_ivp`

The only fortran in `solve_ivp` (the new interface) is `LSODE`, this could either be translated or we could possibly just deprecate it and offer CVODE instead.

Maybe @zachjweiner also has an opinion on this?

**Comment by zachjweiner** [comment]
I started using SciPy for ODE integration after `solve_ivp` was added, so I happily have no ties to the legacy interface and would be glad to see it chucked. I could imagine that there's a userbase that continues to use it, though, since I don't think it's ever been formally declared as deprecated or slated for removal.

However, I'm not sure about dropping/replacing LSODA. As far as I understand it, the A at the end (rather than E) indicates the stiffness detection/automatic switching, and I'm not certain what other methods (e.g., in SUNDIALS) include that feature. (I couldn't find an unambiguous answer on this quickly, so it's worth double checking.) Even still, I'd be a bit careful if/when replacing it because I suspect it sees wide usage. For adaptive methods like ODE solvers (which depend on input data, etc.) I wouldn't be surprised if swapping in a similar (but not identical) replacement leads to surprising breakage in downstream usage.

**Comment by ilayn** [comment]
OK seems like no cavalry on the horizon, back to gotofest then.

**Comment by mdhaber** [comment]
Can we do a one-time release of the legacy interface as a separate package and let it go?

**Comment by ilayn** [comment]
I am a simple man, I see F77 code, I translate. 

If you tell me not to, I will be happy. Not really too crazy about this Don Quixote attempt anyways but I must say, it is embarrassing to have this code in production independent from its functionality.

**Comment by ev-br** [comment]
There's no way to stop Ilhan from translating F77 code. All I'm saying is maybe squint a little and only see the code which is not covered by `solve_ivp` :-).  

From having only looked at all this for the whole of 15 mins --- 
- `solve_ivp` wraps `lsoda` in https://github.com/scipy/scipy/blob/main/scipy/integrate/_ivp/lsoda.py,
- the f2py LSODA wrapper is https://github.com/scipy/scipy/blob/main/scipy/integrate/lsoda.pyf, and
- the F77 beauty is https://github.com/scipy/scipy/blob/main/scipy/integrate/odepack/lsoda.f

**Comment by ilayn** [comment]
No it's the other way around. If you tell me that it is not worth it, then I can skip it. I just want to get this over the line so that we can start working on actual algorithms and not the packaging 24/7. It has been quite sometime that we added some major feature.

**Comment by ev-br** [comment]
Well, LSODA is definitely worth it. The rest, I'm not qualified to say. Am suspecting that not, but it's a guess.

**Comment by mdhaber** [comment]
> what is the user difference between this new package and an old scipy version

Easy - you can have both installed at the same time, so you can use both new features/bug fixes and the legacy functions.

**Comment by ilayn** [comment]
From the look of it, SciPy 2.0 would be an awesome release if it ever comes.

**Comment by dschmitz89** [comment]
Why don't we give https://github.com/sdwfrost/liblsoda a try before rewriting that beast? Looks like a conscious translation to C. FWIW, LSODA seems to be as commonly used for general purpose ODE solving as MINPACK for curve fitting, so it should not be simply removed IMO. There is for example the [numbalsoda](https://github.com/Nicholaswogan/numbalsoda) project for using LSODA from numba which has almost 100 stars.

**Comment by dschmitz89** [comment]
> No it's the other way around. If you tell me that it is not worth it, then I can skip it. Translation gives me no joy you can be sure about that.
> 
> I just want to get this over the line so that we can start working on actual algorithms and not the packaging 24/7. It has been quite sometime that we added some major feature.

Just wanted to mention that we do not need to be that pessimistic in my opinion: despite lots of legacy code, SciPy continues to ship major features. I think for the next release alone we have a completely new integration routine and two new interpolators. I think that is not too bad for a project of our size :).

Your rewriting efforts are still highly appreciated and have already been a major step for `special`!

**Comment by ilayn** [comment]
Ah yes, I meant the features that Fortran code ostensibly offers that were untouchable for a long time. 

But in my defense, I am reading code that says `end of (punch) card 12` at the bottom of the file. So allow some abrasive tone for me :)

**Comment by dschmitz89** [comment]
> Ah yes, I meant the features that Fortran code ostensibly offers that were untouchable for a long time.
> 
> But in my defense, I am reading code that says `end of (punch) card 12` at the bottom of the file. So allow some abrasive tone for me :)

Could you point to the location of that code? Need to include it in any future SciPy talk. ;)

**Comment by ilayn** [comment]
I can't find the `card 12` one but every file in minpack finished with a card statement 

https://github.com/scipy/scipy/blob/f133c27843a36c8fe1c2979b3a072ac4f5a00624/scipy/optimize/minpack/hybrj.f#L438-L440

**Comment by cournape** [comment]
Regarding arpack, I took some times at last scipy conf sprints to take a deeper look at it, including refreshing my old brain on the basic literature behind it.

1. A "clean" implementation of the basics seems easy enough, say a good dedicated week for a first implementation including tests, etc. By basics I mean Hermitian support (Lanczos) with support to find largest eigenvalues. Arnoldi (non Hermitian) should be straightforward extension. Shift-invert, generalized problem, etc. would be more work, not sure by how much.
2. Alternative is to rewrite the fortran code literally in C. See e.g. https://github.com/dgleich/GenericArpack.jl for a julia rewrite almost line by line. That could be done likely more quickly, as you can just do it piece by piece. 

Maybe the pragmatic route is to do 2) first and 1) longer term ?

I can't commit to any timeline, only being interested in doing it w/ some PoC for 1) done.

**Comment by ilayn** [comment]
I agree and (2) is essentially done but I need to pick up the pieces from different machines as I am traveling and then merge. Should be done next week and I'll start testing.

It looks like this in the branch:

https://github.com/ilayn/scipy/tree/rewrite_arpack

**Comment by lucascolley** [comment]
@ilayn we still have `mvnun_weighted` to go before we can tick off the last stats box, right?

**Comment by ilayn** [comment]
I don't mind if it's not ready. I thought we are done finally but it can wait if we are not

**Comment by rkern** [comment]
> * For instance, ODRPACK seems stable and without open issues. So probably low prio.

#23100 seems to be a build issue (though with some flavor of `conda` build, not our wheels), FWIW.

Given the availability of @HugoMVale's [packaging of ODRPACK95 as a standalone package](https://pypi.org/project/odrpack/), I'm more comfortable with deprecating `scipy.odr` entirely now. https://github.com/scipy/scipy/issues/7107#issuecomment-2936918339

I think I'd still welcome a Python/Pythran rewrite of the algorithm, but aimed at `scipy.optimize` and fitting in with its framework.

**Comment by ilayn** [comment]
I feel the same after looking at it for a while now. There is definitely a lot of value to salvage from that package but I don't think it would do any good to distribute it in its current form. Might be much simpler to yank out the algorithm instead, in a more incremental and flexible fashion without the F77-isms forcing us into a particular API.

**Comment by ilayn** [comment]
So after #23963 we are left with ODRPACK which we decided to deprecate and with FITPACK. There is some distance covered already in works like #23962 so it is partially covered. Hence should I get that one out the door?

**Comment by ev-br** [comment]
For the remaing FITPACK item in the OP list, here is a slightly expanded version  of the [discourse message in the ODR thread](https://discuss.scientific-python.org/t/rfc-deprecating-scipy-odr/2166/7):


FITPACK conversion is actually pretty far along. The 1D conversion is done for scipy 1.15, and Gagandeep Singh has just sent a PR with his translation of RectBivariateSpline, https://github.com/scipy/scipy/pull/23962. 

Out of what we actually use from FITPACK, the following routines have no Fortran-free versions:

-  `regrid.f` : gh-23962
-  `surfit.f` : SmoothBivariateSpline / bisplrep -- scattered 2D interpolant

And two spherical routines:

-  `spgrid.f` : RectSpehereBivariateSpline
- `sphere.f` : SmoothSphereBivariateSpline

------------------

One thing we did not converge on for FITPACK is how to pull the plug (that was always for “some far future“).

So far, instead of rewriting old routines with the identical API, we provided functional analogs and declared the FITPACK routines as “legacy”. E.g. splrep and UnivariateSpline are legacy, and the recommended replacement is `make_splrep`.

Now that we can actually plan the removal, the options are:

- make old APIs thin wrappers for over the new routines
- hard remove the old routines and add ample docs on how to port from old to new.

The latter will be quite disruptive indeed. Ralf indicated on Discourse his preference for keeping backwards-compatible wrappers.

(EDIT: comments crossed)

**Comment by lucascolley** [comment]
almost time for the "META: C Code inventory" https://discuss.python.org/t/pre-pep-rust-for-cpython/104906 😄 /s

**Comment by ev-br** [comment]
So @ilayn if you feel like working your magic on FITPACK :-), the largest one is `surfit`. A couple of quick opinionated  notes:
- evaluations should be offloaded to NdBSpline or a subclass, similar to gh-23962
- there are helpers for the knot selection in `interpolate/src/__fitpack.cc`
- the meaty core of the routine is a banded LSQ routine, which sounds right up your alley 
Happy to have a call about details if you're up to it.

**Comment by ilayn** [comment]
I'm now travelling but sure. In the mean time it does not look fun (as if the others did).

<img width="3509" height="2326" alt="Image" src="https://github.com/user-attachments/assets/a0f3780a-1fb5-4183-af56-008fa9bc6cc8" />

**Comment by lorentzenchr** [comment]
Congrats to this achievement!

**Comment by steppi** [comment]
What an accomplishment. Incredible work @ilayn.

**Comment by ilayn** [comment]
![Image](https://github.com/user-attachments/assets/0c0b1654-385f-44dc-93bf-ea78a2c4a4d5)

**Comment by rgommers** [comment]
Amazing work @ilayn! And this went super fast in the end, unbelievable you got it all done before the end of 2025. Can't think of a better Christmas gift than this. 🎁🎄

**Comment by rgommers** [comment]
I'm honestly tempted to poke at removing the Fortran compiler dependency - can't merge that any time soon because of `scipy.odr`, but it does seem like a satisfying thing to do:)

**Comment by ilayn** [comment]
Thanks all for the comments. I hope this leads to some net positive other than just curing our own pains. It was getting desperate towards the end so it was either before Christmas or never. 

And somehow I managed to order them in increasing in difficulty so the last ones were the hardest. The winner in code golfing is hands down bar none LSODA. Nothing beats `stoda.f` so far. It's so clever and at the same time incredibly infuriating.

> I'm honestly tempted to poke at removing the Fortran compiler dependency - can't merge that any time soon because of `scipy.odr` but it does seem like a satisfying thing to do:)

That PR should have `ASMR:` instead of `MAINT:` as the identifier. It would be mesmerizing. I would even try cross-compiling some wheels for platforms we gave up on because of fortran dependency.

**Comment by japui-coder** [comment]
META: FORTRAN Code inventory #18566

**Comment by lucascolley** [comment]
> I'm honestly tempted to poke at removing the Fortran compiler dependency

Me too, https://github.com/lucascolley/scipy/pull/44 seems to do it (delete `scipy.odr` and some `meson.build` stuff).

> It would be mesmerizing.

Here is a log showing no Fortran compiler with a successful build and successful `scipy.linalg` tests[^1]: https://github.com/lucascolley/scipy/actions/runs/20353981370/job/58485246932?pr=44#step:6:63.

[^1]: Forgive me for showing off `pixi global install --path .` :)

**Comment by cournape** [comment]
A wonderful accomplishment !

I still remember Fernando Perez asking around 2010 why we can't just replace the Fortran code and everyone saying it could not happen ! Congrats to the Herculean effort

**Comment by h-vetinari** [comment]
End of an era... Congratulations! 🥳 

> ![Image](https://github.com/user-attachments/assets/0c0b1654-385f-44dc-93bf-ea78a2c4a4d5)

I wasn't aware that insects run Fortran. Save the bees! 😆

**Comment by tupui** [comment]
Massive kudos 👏 And good holiday and new year if you celebrate!

**Comment by Kai-Striega** [comment]
Congratulation! Massive effort from everyone involved.

**Comment by rkern** [comment]
There's something beautiful about a very long checklist, all checked. Well done, and thank you, @ilayn!

**Comment by jjerphan** [comment]
Congratulations and thank you for this massive efforts! 👏

**Comment by anutosh491** [comment]
This is awesome. Thanks for the work here.

Tagging @serge-sans-paille who should also be interested in the same :)

**Comment by serge-sans-paille** [comment]
Well done, /mae carnen/ !

**Comment by rgommers** [comment]
`scipy.odr` removed in gh-25236, so now we're really Fortran-compiler-free.
